src/lib/optimizer/strategy/predicate_reordering_rule.cpp

```diff
@@ -26,114 +20,152 @@ bool is_predicate_style_node(const std::shared_ptr<AbstractLQPNode>& node) {
     return true;
   }
 
-  // Validate can be seen as a Predicate on the MVCC column
+  // Validate can be seen as a Predicate on the MVCC column.
   if (node->type == LQPNodeType::Validate) {
     return true;
   }
 
-  // Semi-/Anti-Joins also reduce the number of tuples and can be freely reordered within a chain of predicates. This
+  // Semi-/anti-joins also reduce the number of tuples and can be freely reordered within a chain of predicates. This
   // might place the join below a ValidateNode, but since it is not a "proper" join (i.e., one that returns columns
-  // from multiple tables), the ValidateNode will still be able to operate on the semi join's output.
+  // from multiple tables), the ValidateNode will still be able to operate on the semi join's output. However, we do not
+  // reorder semi-/anti-joins with multiple join predicates. Though they also filter the relation, they cannot be
+  // executed efficiently.
```

Comment:
So what do you suggest? In fact, we shortcut both in build and probe phase (and of course do not need to randomly access the tuples again for each predicate during probing via `MultiPredicateJoinEvaluator` – random access to unmaterialized column). So guess there's quite a difference regarding efficiency.

Reply:
Well, not taking short cuts that cannot be taken is not inefficient. Same for accessing values that need to be accessed (and we only do it, if there is a (mis-)match and skip otherwise).
So programmatically, it's efficient.

I guess I am fine with something like "multiple predicates can lead to significant overheads and we thus consider them to be more expensive than single-predicate join"?

Reply:
(updated with the comment above)

Reply:
Ack.